### PR TITLE
HAL + SPI updates

### PR DESCRIPTION
**Pull request by virtuald**
I tested these on a RoboRIO before I left, so if the tests pass then I think they're fine unless there's something I accidentally included.

### CODE REVIEW

File: wpilib/wpilib/accumulatorresult.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+#----------------------------------------------------------------------------
+
+from collections import namedtuple
+
+__all__ = ['AccumulatorResult']
+
+AccumulatorResult = namedtuple('AccumulatorResult', ['value', 'count'])
```
Comment (auscompgeek) [inline wpilib/wpilib/accumulatorresult.py]: Hm. Would it make sense to type hint this (using `typing.NamedTuple`)?

File: wpilib/wpilib/spi.py [hunk 2]
```diff
@@ -198,103 +201,331 @@ def read(self, initiate, size):
...
+
+        kAccumulateDepth = 2048
+
+        def __init__(self, port: int, xferSize: int, validMask: int,
+                           validValue: int, dataShift: int, dataSize: int,
+                           isSigned: bool, bigEndian: bool):
+
+            # python-specific: for efficiency purposes, we only support xferSize of
+            #                  2, 4, or 8... if you need to do a different size then
+            #                  file a bug and we'll adjust it
```
Comment (auscompgeek) [inline wpilib/wpilib/spi.py]: Hm. Is using `struct` actually any faster than `int.from_bytes`? I mean I wouldn't be surprised if it is, but I guess numbers would be nice.

File: wpilib/wpilib/spi.py [hunk 3]
```diff
@@ -198,103 +201,331 @@ def read(self, initiate, size):
...
+            elif self._xferSize == 4:
+                self._struct = struct.Struct('%sI' % efmt)
+            elif self._xferSize == 8:
+                self._struct = struct.Struct('%sQ' % efmt)
+            else:
+                raise ValueError("SPI Accumulator only supports xferSize of 2/4/8")
+
+            self._mutex = threading.RLock()
+            self._notifier = Notifier(self._update)
+            self._buf = (ctypes.c_uint8 * (xferSize * self.kAccumulateDepth))()
```
Comment (auscompgeek) [inline wpilib/wpilib/spi.py]: Would using a `bytearray` here work as well?

File: wpilib/wpilib/spi.py [hunk 4]
```diff
@@ -198,103 +201,331 @@ def read(self, initiate, size):
...
-        hal.initSPIAccumulator(self.port, int(period*1.0e6), cmd,
-                               xfer_size, valid_mask, valid_value, data_shift,
-                               data_size, is_signed, big_endian)
+        self.initAuto(xferSize * 2048)
 
-    def freeAccumulator(self):
+        if bigEndian:
+            cmdBytes = cmd.to_bytes(4, 'big')
+        else:
+            cmdBytes = cmd.to_bytes(4, 'little')
```
Comment (auscompgeek) [inline wpilib/wpilib/spi.py]: Tempting to rewrite this as a ternary expression.

File: wpilib/wpilib/accumulatorresult.py (hunk 1 above)
Comment (virtuald) [reply wpilib/wpilib/accumulatorresult.py]: Knock yourself out.

File: wpilib/wpilib/spi.py (hunk 3 above)
Comment (virtuald) [reply wpilib/wpilib/spi.py]: I don't know.

File: wpilib/wpilib/spi.py (hunk 2 above)
Comment (virtuald) [reply wpilib/wpilib/spi.py]: well, from_bytes doesn't support a buffer offset? I think.

Comment (auscompgeek) [reply wpilib/wpilib/spi.py]: Ah, good point; I missed that part. You could definitely use a `memoryview` to do that though.

**Comment by auscompgeek** [comment]
Say, was spi.py missing an updated git-source-track line?

**Pull request merged**